torch/nn/modules/module.py

```diff
@@ -802,8 +802,13 @@ def compute_should_use_set_data(tensor, tensor_applied):
             with torch.no_grad():
                 param_applied = fn(param)
             p_should_use_set_data = compute_should_use_set_data(param, param_applied)
+
+            # subclasses may have multiple child tensors so we need to use swap_tensors
+            is_subclass = isinstance(param_applied, torch.Tensor) and type(param_applied) is not torch.Tensor
```

Comment:
cc @albanD @mikaylagawarecki we are enabling this by default for tensor subclasses, wondering if this make sense or we should only enable this when loading the dtensor module?

Reply:
I wonder if there exists non wrapper subclasses where we would still want this behavior, probably not since we already have the future flag

Reply:
Do we have a full list of subclasses (wrapper and non wrapper)? I can add some more tests/sanity check this

Reply:
A safer way to check is to utilize this [check](https://github.com/pytorch/pytorch/blob/main/torch/utils/_python_dispatch.py#L254) function, where it only apply for "wrapper" subclass that is well defined already (i.e. DTensor, Float8Tensor)

Reply:
Making this the default for wrapper subclasses sounds reasonable to me and was on my list. I don't think we should make this the default for non-wrapper subclasses yet.

Reply:
Could you elaborate more on what you mean by it doesn't work with `Parameter`?

`torch.nn.utils.swap_tensors` does have certain constraints

- tensors cannot have python weakrefs to them
- `__slots__` of tensors swapped must match
- `use_count` and `weak_use_count` of `THPVariable->cdata` must be 1

https://github.com/pytorch/pytorch/blob/5af839f86d3be4d3299fe85b46cdbaa146e5c992/torch/utils/__init__.py#L34-L42

https://github.com/pytorch/pytorch/blob/5af839f86d3be4d3299fe85b46cdbaa146e5c992/torch/csrc/Module.cpp#L370-L386

The issue with RNN/GRU/LSTM that you might have seen in the failures is that these modules hold weakrefs to some of the params, and is a known issue.  #122800 might be a fix for this.

Reply:
I'm not sure the exact issue but seems like .to is being called on nn.Parameter (thus passed my initial subclass check)

Parameter causing issues is LayerNorm.weight -- `RuntimeError: Expected single reference to a's Tensor object but got 3`

```
<class 'torch.nn.parameter.Parameter'>
E
======================================================================
ERROR: test_transformerencoder_batch_first_False_training_True_enable_nested_tensor_False_cpu (__main__.TestTransformersCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/nn/modules/module.py", line 819, in _apply
    torch.utils.swap_tensors(param, param_applied)
  File "/home/tristanr/pytorch/torch/utils/__init__.py", line 68, in swap_tensors
    torch._C._swap_tensor_impl(t1, t2)
RuntimeError: Expected single reference to a's Tensor object but got 3

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/testing/_internal/common_utils.py", line 2741, in wrapper
    method(*args, **kwargs)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_device_type.py", line 432, in instantiated_test
    raise rte
  File "/home/tristanr/pytorch/torch/testing/_internal/common_device_type.py", line 419, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/tristanr/pytorch/test/test_transformers.py", line 643, in test_transformerencoder
    _test(batch_first, training, enable_nested_tensor)
  File "/home/tristanr/pytorch/test/test_transformers.py", line 615, in _test
    enable_nested_tensor=enable_nested_tensor).to(device)
  File "/home/tristanr/pytorch/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
  File "/home/tristanr/pytorch/torch/nn/modules/module.py", line 778, in _apply
    module._apply(fn)
  File "/home/tristanr/pytorch/torch/nn/modules/module.py", line 823, in _apply
    raise RuntimeError(f"_apply(): Couldn't swap {self._get_name()}.{key}") from e
RuntimeError: _apply(): Couldn't swap LayerNorm.weight

To execute this test, run the following from the base repo dir:
     python test/test_transformers.py -k test_transformerencoder_batch_first_False_training_True_enable_nested_tensor_False_cpu

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

Reply:
So seems to be a refcount issue -- not sure why LayerNorm.weight has multiple refs

Reply:
Hm, I have tested the behavior of `swap_tensors` with `_apply` on all nn.Modules in [these tests](https://github.com/pytorch/pytorch/blob/main/test/test_modules.py?fbclid=IwAR3ytpr8RQjdgcxIW_z5sMEOaAEcTftAYDOHwqWiiXBbV8oIMiS2K_pDNtA#L866-L983) and the only ones with issues are RNN/GRU/LSTM.

So I do not think it is an issue with the LayerNorm/TransformerEncoder module itself, though perhaps there is some logic in the test in question that causes this

Reply:
Filed https://github.com/pytorch/pytorch/issues/122803 to investigate this